### PR TITLE
feat(otel): metrics support

### PR DESCRIPTION
**Pull request by mabdinur**
## What does this PR do?

Adds OpenTelemetry Metrics SDK support with OTLP transport to `datadog-opentelemetry`. Metrics can be exported via gRPC (default) or HTTP/protobuf, enabling Rust applications to export OpenTelemetry metrics to Datadog using the standard OpenTelemetry Metrics API.

### Key Features

- **OTLP Transport**: Supports gRPC (default, port 4317) and HTTP/protobuf (port 4318) protocols
- **Configuration API**: Full support for OTLP metrics configuration via `ConfigBuilder`, including `set_otel_resource_attributes()`
- **Resource Attributes**: Support for `OTEL_RESOURCE_ATTRIBUTES` environment variable
- **Temporality Control**: Configurable Delta/Cumulative temporality preference
- **Graceful Degradation**: HTTP/JSON protocol gracefully degrades with a warning (no-op provider)

## Motivation

Enables Rust applications to export OpenTelemetry metrics to Datadog using the standard OpenTelemetry Metrics API, aligning with support in other languages and providing a consistent observability experience across the Datadog ecosystem.

## Additional Notes

- **Metrics are opt-in** via `DD_METRICS_OTEL_ENABLED=true`

**Feature gating:**
- **Default**: `cargo add datadog-opentelemetry` → gRPC transport
- **HTTP-only**: `cargo add datadog-opentelemetry --no-default-features --features metrics-http` → HTTP/protobuf without gRPC dependencies
- **Both**: `cargo add datadog-opentelemetry --features metrics-grpc,metrics-http` → enables both protocols

**Testing:**
- Implementation is fully tested here: https://github.com/DataDog/system-tests/pull/5863

### CODE REVIEW

**Comment by mabdinur** [comment]
Shared tests covering edge cases: https://github.com/DataDog/system-tests/pull/5863

File: datadog-opentelemetry/src/core/configuration/configuration.rs [hunk 1]
```diff
@@ -1011,6 +1077,16 @@ impl Config {
             &self.remote_config_enabled,
             &self.remote_config_poll_interval,
             &self.datadog_tags_max_length,
+            &self.otlp_endpoint,
```
Comment (iunanua) [inline datadog-opentelemetry/src/core/configuration/configuration.rs]: is metrics_otel_enabled missing?
```suggestion
            &self.metrics_otel_enabled,
            &self.otlp_endpoint,
```

File: datadog-opentelemetry/src/lib.rs [hunk 2]
```diff
@@ -477,3 +513,54 @@ pub fn make_test_tracer(shared_config: Arc<Config>) -> (SdkTracerProvider, Datad
...
+        self
+    }
+
+    pub fn init(self) -> Result<SdkMeterProvider, String> {
+        let (meter_provider, _) = self.init_local()?;
+        opentelemetry::global::set_meter_provider(meter_provider.clone());
+        Ok(meter_provider)
+    }
+
+    pub fn init_local(self) -> Result<(SdkMeterProvider, ()), String> {
```
Comment (iunanua) [inline datadog-opentelemetry/src/lib.rs]: Is the tuple needed?
```suggestion
    pub fn init_local(self) -> Result<SdkMeterProvider, String> {
```

File: datadog-opentelemetry/Cargo.toml [hunk 3]
```diff
@@ -76,7 +79,10 @@ tracing-subscriber = { version = "0.3" }
 criterion = "0.5.1"
 
 [features]
-test-utils = ["libdd-data-pipeline/test-utils", "criterion"]
+default = ["metrics-grpc"]
```
Comment (iunanua) [inline datadog-opentelemetry/Cargo.toml]: I think we didn't want default features, wdyt @bantonsson?

File: Cargo.lock
Comment (paullegranddc) [inline Cargo.lock]: The Cargo lock file is somehow not up to date when I pull this branch.
You probably need to do a `cargo check` and commit the resulting lockfile

File: datadog-opentelemetry/src/lib.rs [hunk 4]
```diff
@@ -477,3 +513,54 @@ pub fn make_test_tracer(shared_config: Arc<Config>) -> (SdkTracerProvider, Datad
...
+        let config = self.config.unwrap_or_else(|| Config::builder().build());
+        let meter_provider = crate::metrics_reader::create_meter_provider(
+            Arc::new(config),
+            self.resource,
+            self.export_interval,
+        )?;
+        Ok((meter_provider, ()))
+    }
+}
```
Comment (paullegranddc) [inline datadog-opentelemetry/src/lib.rs]: I don't think those methods should ever return errors.
I'd prefer if we returned an no-op SdkMeterProvider (with no exporter) and logged a scary log, rather than customers having to think about what if the meter provider is missing.

Also in general, return `Result<T, String>` feels weird... I don't think it's completely invalid, but I'd prefer if you used `anyhow::Result<T>` instead.
The error is also typed erased and can be a just a string, but it also captures the backtrace and you can wrap errors more efficiently.

File: datadog-opentelemetry/src/metrics_exporter.rs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+pub fn get_otlp_metrics_timeout(config: &Config) -> u32 {
+    let timeout = config.otlp_metrics_timeout();
+    if timeout != 0 {
+        return timeout;
+    }
+    config.otlp_timeout()
+}
+
+pub fn get_metric_export_interval_ms(config: &Config) -> u32 {
+    config.metric_export_interval()
```
Comment (paullegranddc) [inline datadog-opentelemetry/src/metrics_exporter.rs]: A single line function not adding anything?

File: datadog-opentelemetry/src/lib.rs [hunk 6]
```diff
@@ -165,10 +198,13 @@ pub(crate) mod propagation;
 pub(crate) mod sampling;
 
 mod ddtrace_transform;
+pub mod metrics_exporter;
+pub mod metrics_reader;
```
Comment (paullegranddc) [inline datadog-opentelemetry/src/lib.rs]: Why are these modules publicly exposed?
The only item that  should probably be public is the `OtlpProtocol` enum, if we expose it in the config

File: datadog-opentelemetry/src/core/configuration/configuration.rs [hunk 7]
```diff
@@ -780,6 +780,12 @@ pub struct Config {
     /// A list of default tags to be added to every span
     /// If DD_ENV or DD_VERSION is used, it overrides any env or version tag defined in DD_TAGS
     global_tags: ConfigItem<Vec<(String, String)>>,
+    /// OTEL resource attributes parsed from OTEL_RESOURCE_ATTRIBUTES env var
+    otel_resource_attributes: ConfigItem<Vec<(String, String)>>,
+    /// OTEL metrics exporter type
+    otel_metrics_exporter: ConfigItem<Cow<'static, str>>,
+    /// OTEL metrics temporality preference
+    otel_metrics_temporality_preference: ConfigItem<Cow<'static, str>>,
```
Comment (paullegranddc) [inline datadog-opentelemetry/src/core/configuration/configuration.rs]: same remark about type conversion with the temporality preference, I think it's better to do validation and cast to an enum here.

The manual `set_otel_metrics_temporality_preference` method should take an enum type instead of a string for instance

File: datadog-opentelemetry/src/metrics_reader.rs [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+            Duration::from_millis(get_metric_export_interval_ms(&config) as u64)
+        });
+
+        let telemetry_exporter = TelemetryTrackingExporter::new(exporter, protocol);
+
+        let reader = opentelemetry_sdk::metrics::PeriodicReader::builder(telemetry_exporter)
+            .with_interval(interval)
+            .build();
+
+        let mut resource_attrs: Vec<opentelemetry::KeyValue> = Vec::new();
```
Comment (paullegranddc) [inline datadog-opentelemetry/src/metrics_reader.rs]: Could you isolate all dd config -> otel resource manipulation to a separate function? I think it'd be easier to read

File: datadog-opentelemetry/src/telemetry_metrics_exporter.rs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                OtelMetricsExportFailuresGrpc,
+            ),
+            OtlpProtocol::HttpProtobuf | OtlpProtocol::HttpJson => (
+                OtelMetricsExportAttemptsHttp,
+                OtelMetricsExportSuccessesHttp,
+                OtelMetricsExportFailuresHttp,
+            ),
+        };
+
+        telemetry::add_point(1.0, attempts);
```
Comment (paullegranddc) [inline datadog-opentelemetry/src/telemetry_metrics_exporter.rs]: Is it fine to add one telemetry point per point exported? Or maybe exports are not that frequent?

File: datadog-opentelemetry/tests/integration_tests/metrics.rs [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+            result.observe(15, &[KeyValue::new("key6", "value6")]);
+        })
+        .build();
+}
+
+#[tokio::test]
+async fn test_metrics_export_grpc() {
+    const SESSION_NAME: &str = "opentelemetry_api/test_metrics_grpc";
+
+    let test_agent = setup_test_agent(SESSION_NAME).await;
```
Comment (paullegranddc) [inline datadog-opentelemetry/tests/integration_tests/metrics.rs]: The test agent version/client we use doesn't have the otlp endppints yet.
The support has been merged in libdatadog but we need to bump it here (we won't need a new release, just update to a git hash probably)

File: datadog-opentelemetry/src/lib.rs (hunk 6 above)
Comment (mabdinur) [reply datadog-opentelemetry/src/lib.rs]: Good catch. I will make these modules internal to the crate. They should not be exposed to users

File: datadog-opentelemetry/src/core/configuration/configuration.rs (hunk 7 above)
Comment (mabdinur) [reply datadog-opentelemetry/src/core/configuration/configuration.rs]: We were actually missing the `set_otel_metrics_temporality_preference` method. I added it to the programatic api for configurations.

File: datadog-opentelemetry/src/metrics_reader.rs (hunk 8 above)
Comment (mabdinur) [reply datadog-opentelemetry/src/metrics_reader.rs]: I moved this logic to it's own method but I kept this logic in the metric reader (since this is the only place it's currently used). 

In a follow up PR I can move this logic to it's own module. Datadog SDKs should support the `OTEL_RESOURCE_ATTRIBUTES` configuration for tracing (and eventually logs). Here's a doc with all the configurations we need to support: https://docs.google.com/document/d/1gtdjhsMSIvuTnM9E4EreI7oZ-bAUoAxkhXX0-fBvdDY/edit?tab=t.0#heading=h.rnd972k0hiye

File: datadog-opentelemetry/src/telemetry_metrics_exporter.rs (hunk 9 above)
Comment (mabdinur) [reply datadog-opentelemetry/src/telemetry_metrics_exporter.rs]: We are adding one telemetry point per export. By default we will export metrics every 10 seconds, so this should be fine for most use cases.

File: datadog-opentelemetry/src/metrics_exporter.rs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+    let endpoint = config.otlp_endpoint();
+    if !endpoint.is_empty() {
+        return Ok(endpoint.to_string());
+    }
+
+    let agent_url = config.trace_agent_url();
+    let url = agent_url
+        .parse::<hyper::http::Uri>()
+        .map_err(|e| format!("Invalid agent URL: {e}"))?;
```
[the remark this answers is not in the document]
Comment (mabdinur) [reply datadog-opentelemetry/src/metrics_exporter.rs]: Nope. I updated this logic. At this point we can assume we have a valid url.
